### PR TITLE
CR-1151706 XRT service bundle for collecting analytics data

### PR DESCRIPTION
**Pull request by xdavidz**
Signed-off-by: David Zhang <davidzha@xilinx.com>

#### Problem solved by the commit

More work needed to make sure this script works across different OS versions.
example of a typical service bundle

```
# /proj/rdi/staff/davidzha/XRT_ssh/src/runtime_src/tools/scripts/service_bundle.sh -d d9
card: d9
use configured XRT
dump into: ./xrt_bundle
running: host_info - progress: 1/6.
running: xbmgmt_output - progress: 2/6.
running: xbutil_output - progress: 3/6.
running: vmr_logs - progress: 4/6.
running: debug_logs - progress: 5/6.
running: core_dumps - progress: 6/6 - skipped: -a to enable.
packaging data...
Estimated: [==================================================]
Progress:  [>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>]
/root/xrt_bundle.tar.gz
11M

[root@xsjyliu51 ~]# tree xrt_bundle
xrt_bundle
├── core_dumps
├── debug_logs
│   ├── dmesg_current.txt
│   ├── sys_kernel_debug_xclmgmt.txt
│   ├── sys_kernel_debug_xocl.txt
│   ├── var
│   │   └── log
│   │       ├── anaconda
│   │       │   ├── anaconda.log
│   │       │   ├── dbus.log
│   │       │   ├── dnf.librepo.log
[trimmed]
│   ├── xclmgmt_parameters_health_check.txt
│   ├── xclmgmt_parameters_health_interval.txt
│   ├── xclmgmt_parameters_mailbox_no_intr.txt
│   ├── xclmgmt_parameters_mailbox_test_mode.txt
│   ├── xclmgmt_parameters_minimum_initialization.txt
│   ├── xclmgmt_parameters_rom_uuid.txt
│   ├── xclmgmt_parameters_vmr_sc_ready_timeout.txt
│   ├── xclmgmt_parameters_xrt_debug_bufsize.txt
│   ├── xclmgmt_parameters_xrt_reset_syncup.txt
│   ├── xocl_parameters_desc_blen_max.txt
│   ├── xocl_parameters_desc_set_depth.txt
│   ├── xocl_parameters_enable_credit_mp.txt
│   ├── xocl_parameters_interrupt_mode.txt
│   ├── xocl_parameters_mailbox_no_intr.txt
│   ├── xocl_parameters_mailbox_test_mode.txt
│   ├── xocl_parameters_p2p_max_bar_size.txt
│   ├── xocl_parameters_poll_mode.txt
│   ├── xocl_parameters_qdma_interrupt_mode.txt
│   ├── xocl_parameters_qdma_max_channel.txt
│   ├── xocl_parameters_rom_uuid.txt
│   ├── xocl_parameters_xrt_debug_bufsize.txt
│   └── xocl_parameters_xrt_reset_syncup.txt
├── host_info
│   ├── lsblk.txt
│   ├── lsb_release_d.txt
│   ├── lsb_release.txt
│   ├── lscpu.txt
│   ├── lsmem.txt
│   ├── lspci.txt
│   ├── lspci_vvv.txt
│   ├── proc_interrupts.txt
│   ├── proc_iomem.txt
│   ├── top.txt
│   ├── uname_a.txt
│   └── vmstat_s.txt
├── vmr_logs
│   ├── vmr_endpoint
│   ├── vmr_log
│   ├── vmr_mem_stats
│   ├── vmr_plm_log
│   ├── vmr_status
│   ├── vmr_system_dtb
│   ├── vmr_task_stats
│   ├── vmr_verbose_info
│   └── xbmgmt_vmr_verbose.txt
├── xbmgmt_output
│   ├── xbmgmt_examine_all.txt
│   └── xbmgmt_examine.txt
└── xbutil_output
    ├── xbutil_examine_all.txt
    ├── xbutil_examine.txt
    └── xbutil_validate.txt
```
example of a run with core dump
```
# /proj/rdi/staff/davidzha/XRT_ssh/src/runtime_src/tools/scripts/service_bundle.sh -d 81 -o /root/test1 -a
card: 81
XRT is not found, use default XRT
dump into: /root/test1/xrt_bundle
running: host_info - progress: 1/6.
running: xbmgmt_output - progress: 2/6.
running: xbutil_output - progress: 3/6.
running: vmr_logs - progress: 4/6.
running: debug_logs - progress: 5/6.
running: core_dumps - progress: 6/6.
packaging data...
Estimated: [==================================================]
Progress:  [>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>]
/root/test1/xrt_bundle.tar.gz
340M

oot@xsjdavidz01:~# tree test1
test1
├── xrt_bundle
│   ├── core_dumps
│   │   ├── cores
│   │   │   └── test.core
│   │   └── var
│   │       └── crash
│   │           └── _opt_xilinx_xrt_test_validate.exe.0.crash
│   ├── debug_logs
│   │   ├── dmesg_current.txt
│   │   ├── sys_kernel_debug_xclmgmt.txt
│   │   ├── sys_kernel_debug_xocl.txt

[trimmed]
```

### CODE REVIEW

**Review by stsoe — APPROVED** [review]

**Review by maxzhen — APPROVED** [review]
Do we want to also collect core dumps (kernel or user process), if there is one?

**Comment by xdavidz** [comment]
> Do we want to also collect core dumps (kernel or user process), if there is one?

added an extra option -a for collecting core dumps, because those may be huge on some systems.

**Review by larry9523 — APPROVED** [review]

File: src/runtime_src/tools/scripts/service_bundle.sh [hunk 1]
```diff
@@ -0,0 +1,236 @@
+#!/bin/bash
+
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (C) 2019-2021 Xilinx, Inc. All rights reserved.
```
Comment (rozumx) [inline src/runtime_src/tools/scripts/service_bundle.sh:4]: Hmmm....Is this a new file or was it moved.   If new, then the copyright date needs to be corrected.

File: src/runtime_src/tools/scripts/service_bundle.sh [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+# This script collects debugging logs for XRT managed sub-systems
+
+TIME=`date +%F-%T`
+ID=`id -un`
+FULL_CMD=${BASH_SOURCE[0]}
+BASE_CMD=`basename "$0"`
+
+progress() {
+	echo -n "running: $func - progress: $(($1+1))/$2"
+	#echo -ne "running: $func - progress: \b\b\b$(($1+1))/$(($2+1))\b\b\b"
```
Comment (rozumx) [inline src/runtime_src/tools/scripts/service_bundle.sh:15]: There shouldn't be any dead code.  

If this functionality is needed, then an helper function should be created and depending on an option or environment variable then the alternate message should be produced.

File: src/runtime_src/tools/scripts/service_bundle.sh [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+		cat /sys/module/xocl/parameters/$xocl > $1/xocl_parameters_$xocl.txt
+	done
+
+	dmesg -T > $1/dmesg_current.txt
+
+	cat /sys/kernel/debug/xclmgmt/trace > $1/sys_kernel_debug_xclmgmt.txt &
+	PID1=$!
+	cat /sys/kernel/debug/xocl/trace > $1/sys_kernel_debug_xocl.txt &
+	PID2=$!
+	sleep 2
```
Comment (rozumx) [inline src/runtime_src/tools/scripts/service_bundle.sh:118]: Why the magic sleep.  A comment here would be nice to indicate what is being waited for.

File: src/runtime_src/tools/scripts/service_bundle.sh [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+fi
+
+if [ -z $PATH_TO_OUTPUT ];then
+	PATH_TO_OUTPUT="."
+fi
+
+PATH_TO_BUNDLE=$PATH_TO_OUTPUT/xrt_bundle
+echo "dump into: $PATH_TO_BUNDLE"
+
+#a=${!funcArray[@]}
```
Comment (rozumx) [inline src/runtime_src/tools/scripts/service_bundle.sh:207]: Please remove dead code.

File: src/runtime_src/tools/scripts/service_bundle.sh [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+# Copyright (C) 2019-2021 Xilinx, Inc. All rights reserved.
+#
+# This script collects debugging logs for XRT managed sub-systems
+
+TIME=`date +%F-%T`
+ID=`id -un`
+FULL_CMD=${BASH_SOURCE[0]}
+BASE_CMD=`basename "$0"`
+
+progress() {
```
Comment (rozumx) [inline src/runtime_src/tools/scripts/service_bundle.sh:13]: To help future developers, having a comment here describing what each argument is would be helpful, for example:
`# $1 - Index; $2 - Total; $3 - Function`

**Review by rozumx — APPROVED** [review]
Overall, the functionality of this bash script looks good.  

To improve upon it, it would be good some if some comments were modified and/or added.  For example:
* Copyright date
* Dead code removal
* Description of arguments being passed into functions (e.g., what each place holder is)

File: src/runtime_src/tools/scripts/service_bundle.sh (hunk 1 above)
Comment (xdavidz) [reply src/runtime_src/tools/scripts/service_bundle.sh]: fixed.

File: src/runtime_src/tools/scripts/service_bundle.sh (hunk 2 above)
Comment (xdavidz) [reply src/runtime_src/tools/scripts/service_bundle.sh]: removed

File: src/runtime_src/tools/scripts/service_bundle.sh (hunk 3 above)
Comment (xdavidz) [reply src/runtime_src/tools/scripts/service_bundle.sh:123]: sometimes the cat might takes about 1 seconds if there are too many data in the debugfs. I think 2 seconds is safer.

File: src/runtime_src/tools/scripts/service_bundle.sh (hunk 4 above)
Comment (xdavidz) [reply src/runtime_src/tools/scripts/service_bundle.sh:217]: will fix it

File: src/runtime_src/tools/scripts/service_bundle.sh (hunk 5 above)
Comment (xdavidz) [reply src/runtime_src/tools/scripts/service_bundle.sh:14]: fixed

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Pull request merged**